File: specs/typespec/audit-logs/audit-logs.go
package auditlogs

import (
	"fmt"
	"time"

	"vetchium-api-server.typespec/common"
)

const (
	defaultAuditLogLimit = 40
	maxAuditLogLimit     = 100
	minAuditLogLimit     = 1

	errAuditLogLimitInvalid     = "must be between 1 and 100"
	errAuditLogStartTimeInvalid = "must be a valid ISO 8601 timestamp"
	errAuditLogEndTimeInvalid   = "must be a valid ISO 8601 timestamp"
)

// AuditLogEntry is a single audit log record returned by the filter APIs.
type AuditLogEntry struct {
	ID           string                 `json:"id"`
	EventType    string                 `json:"event_type"`
	ActorUserID  *string                `json:"actor_user_id"`
	TargetUserID *string                `json:"target_user_id"`
	OrgID        *string                `json:"org_id"`
	IPAddress    string                 `json:"ip_address"`
	EventData    map[string]interface{} `json:"event_data"`
	CreatedAt    string                 `json:"created_at"`
}

// FilterAuditLogsRequest is the shared request body for all filter-audit-logs endpoints.
type FilterAuditLogsRequest struct {
	EventTypes    []string `json:"event_types,omitempty"`
	ActorUserID   *string  `json:"actor_user_id,omitempty"`
	StartTime     *string  `json:"start_time,omitempty"`
	EndTime       *string  `json:"end_time,omitempty"`
	PaginationKey *string  `json:"pagination_key,omitempty"`
	Limit         *int32   `json:"limit,omitempty"`
}

func (r FilterAuditLogsRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Limit != nil {
		if *r.Limit < minAuditLogLimit || *r.Limit > maxAuditLogLimit {
			errs = append(errs, common.NewValidationError("limit", fmt.Errorf(errAuditLogLimitInvalid)))
		}
	}

	if r.StartTime != nil {
		if _, err := time.Parse(time.RFC3339, *r.StartTime); err != nil {
			errs = append(errs, common.NewValidationError("start_time", fmt.Errorf(errAuditLogStartTimeInvalid)))
		}
	}

	if r.EndTime != nil {
		if _, err := time.Parse(time.RFC3339, *r.EndTime); err != nil {
			errs = append(errs, common.NewValidationError("end_time", fmt.Errorf(errAuditLogEndTimeInvalid)))
		}
	}

	return errs
}

// FilterAuditLogsResponse is the response for all filter-audit-logs endpoints.
type FilterAuditLogsResponse struct {
	AuditLogs     []AuditLogEntry `json:"audit_logs"`
	PaginationKey *string         `json:"pagination_key"`
}

// EffectiveLimit returns the limit to use for a query, applying the default if none specified.
func (r FilterAuditLogsRequest) EffectiveLimit() int32 {
	if r.Limit != nil {
		return *r.Limit
	}
	return defaultAuditLogLimit
}

File: specs/typespec/audit-logs/audit-logs.ts
import { newValidationError, type ValidationError } from "../common/common";

export interface AuditLogEntry {
	id: string;
	event_type: string;
	actor_user_id: string | null;
	target_user_id: string | null;
	org_id: string | null;
	ip_address: string;
	event_data: Record<string, unknown>;
	created_at: string; // ISO 8601
}

export interface FilterAuditLogsRequest {
	event_types?: string[];
	actor_user_id?: string;
	start_time?: string; // ISO 8601
	end_time?: string; // ISO 8601
	pagination_key?: string;
	limit?: number; // 1-100, default 40
}

export interface FilterAuditLogsResponse {
	audit_logs: AuditLogEntry[];
	pagination_key: string | null;
}

export function validateFilterAuditLogsRequest(
	request: FilterAuditLogsRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (request.limit !== undefined) {
		if (!Number.isInteger(request.limit) || request.limit < 1 || request.limit > 100) {
			errs.push(newValidationError("limit", "must be between 1 and 100"));
		}
	}

	if (request.start_time !== undefined) {
		const d = new Date(request.start_time);
		if (isNaN(d.getTime())) {
			errs.push(newValidationError("start_time", "must be a valid ISO 8601 timestamp"));
		}
	}

	if (request.end_time !== undefined) {
		const d = new Date(request.end_time);
		if (isNaN(d.getTime())) {
			errs.push(newValidationError("end_time", "must be a valid ISO 8601 timestamp"));
		}
	}

	return errs;
}
